File: 0496-next-greater-element-i/0496-next-greater-element-i.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        unordered_map<int, int> nums2_table;
        int length1 = nums1.size(), length2 = nums2.size(); 
        vector<int> ans;
        for (int i = 0; i < length2; ++i) nums2_table[nums2[i]] = i;
        
        for (int i = 0; i < length1; ++i){
            try {
                for (int j = nums2_table[nums1[i]]; j < length2; ++j){
                    if (nums1[i] < nums2[j]) {ans.push_back(nums2[j]); break;}
                    if (j == length2-1) ans.push_back(-1);
                } 
            }
            
            catch(...){
                ans.push_back(-1);
            }
        }
        
        return ans;
    }
};

File: 0496-next-greater-element-i/0496-next-greater-element-i.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        stack, m, len1, len2 = [], dict(), len(nums1), len(nums2)
        
        for i in range(0, len2):
            while (stack and nums2[i] > stack[-1]):
                m[stack.pop()] = nums2[i]
            stack.append(nums2[i])
        
        while stack:
            m[stack.pop()] = -1
        
        ans = [0] * len1
        
        for i in range(0, len1):
            ans[i] = m[nums1[i]]
        return ans
        
                